### PR TITLE
User view stuff

### PR DESCRIPTION
**Pull request by KevinDeMars**

### CODE REVIEW

File: food-truck-api/src/main/java/food/truck/api/user/AbstractUser.java [hunk 1]
```diff
@@ -8,10 +8,11 @@
 @Data
 @AllArgsConstructor
 @NoArgsConstructor
-public class AbstractUser {
+public abstract class AbstractUser {
     protected Position position;
 
     public void visit(UserVisitor v) {
```
Comment (MicahSchiewe) [inline food-truck-api/src/main/java/food/truck/api/user/AbstractUser.java:14]: I think we can delete this method now

File: food-truck-frontend/src/api/RateReviewApi.ts [hunk 2]
```diff
@@ -11,37 +13,33 @@ import {ClientResponseAction} from "./ResponseTypes";
 export const loadReviewsByUser = async (
   userId: number,
   onFail?: ClientResponseAction
-) => {
-  let reviews: Review[] = [];
+): Promise<Review[] | null> => {
```
Comment (MicahSchiewe) [inline food-truck-frontend/src/api/RateReviewApi.ts:16]: If it returns async, it implicitly returns a promise of its specified return type. So I'd expect this to return `Promise<Promise<Review[] | null>>`. But I could be wrong.

File: food-truck-frontend/src/api/RateReviewApi.ts [hunk 3]
```diff
@@ -95,59 +86,46 @@ export const loadReviewFromTruckForUser = async (
...
+      .filter(r => r != null) as Review[];
 
-    // Fetch the truck's info. Yes, I could've gotten it from the reviews. That seemed like a coupling hazard, what happens
-    // if we change the way reviews get stored and they suddenly don't store trucks directly?
-    let truck: Truck | null = truck_backendToFrontend((await api.request({
-      url: `/truck/${truckId}`,
-      method: "GET"
-    })).data);
-    console.log(truck);
+    const truck = await getTruckById(truckId, () => {
```
Comment (MicahSchiewe) [inline food-truck-frontend/src/api/RateReviewApi.ts:113]: Is there a reason we're swallowing the error here?

File: food-truck-frontend/src/api/RouteLocationApi.ts [hunk 4]
```diff
@@ -1,53 +1,38 @@
...
-    })
-    .then((response) => {
-      if (response.data)
-        routePts = response.data.locations.map((pt: any) =>
-          backendToFrontend(pt, nextStopId++)
-        );
-    })
-    .catch(onFail);
-  return routePts;
+  const resp = await api.get(`/truck/${truckId}/active-route`);
```
Comment (MicahSchiewe) [inline food-truck-frontend/src/api/RouteLocationApi.ts:13]: So we're no longer passing the current date/time? Was the backend updated accordingly?

File: food-truck-frontend/src/api/SubscriptionApi.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+export const subscribeToTruck = async (truckId: number): Promise<Subscription | null> => {
+  let res = await api.post(`/truck/${truckId}/subscribe`);
+  if (res.data == null)
+    return null;
+  return parse(SubscriptionMeta, res.data);
+}
+
+export const unsubscribeFromTruck = async (truckId: number) => {
+  await api.delete(`/truck/${truckId}/unsubscribe`);
```
Comment (MicahSchiewe) [inline food-truck-frontend/src/api/SubscriptionApi.ts:25]: I guess unsubscribing provides no feedback regarding success?

File: food-truck-frontend/src/components/Form.tsx [hunk 6]
```diff
@@ -35,48 +36,67 @@ class Form extends Component<Props, State> {
...
+      if (e.props.name) {
+        this.state.formData[e.props.name] =
+            e.props.value ??
+            e.props.defaultValue ??
+            e.props.checked ??
+            "";
       }
     });
 
+    //console.log("Initial state: ");
```
Comment (MicahSchiewe) [inline food-truck-frontend/src/components/Form.tsx:53]: If we have time, we should probably delete commented-out code. But I'm not sure we have time.

File: food-truck-frontend/src/components/Form.tsx [hunk 7]
```diff
@@ -35,48 +36,67 @@ class Form extends Component<Props, State> {
...
-          onChange: this.onValueChanged,
-          value: this.state.formData[c.props.name],
-          defaultValue: undefined
-        })}
-      </Grid>
-    );
+
+    const c = child as ReactElement;
+    if (c.props.name) {
+      //console.log("Mapping " + c.props.name + " => " + this.state.formData[c.props.name]);
```
Comment (MicahSchiewe) [inline food-truck-frontend/src/components/Form.tsx:74]: Commented-out code

File: food-truck-frontend/src/components/Form.tsx [hunk 8]
```diff
@@ -123,25 +143,34 @@ class Form extends Component<Props, State> {
...
-        ...prev.formData,
-        [name]: value
-      }
-    }));
+        formData: {
+          ...prev.formData,
+          [name]: value
+        }
+      }),
+//() => console.log(this.state)
```
Comment (MicahSchiewe) [inline food-truck-frontend/src/components/Form.tsx:151]: Commented-out code

File: food-truck-frontend/src/components/SubscriptionList.tsx [hunk 9]
```diff
@@ -66,19 +38,12 @@ class SubscriptionList extends React.Component<SubscriptionProps, SubscriptionSt
                                     <tr>
                                             <th>Food Trucks</th>
                                     </tr>
-                                    {this.state.data.map((value, index) => this.renderSubscriptionElement(index))}
```
Comment (MicahSchiewe) [inline food-truck-frontend/src/components/SubscriptionList.tsx:69]: Where is the replacement rendering subscriptions?

File: food-truck-frontend/src/components/truck/rate_and_review/ReviewView.tsx [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+type Props = {
+    review: Review
+}
+
+class ReviewView extends React.Component<Props, {}> {
+    constructor(props: Props, context: {}) {
+        super(props, context);
+    }
+    
+    /*componentDidMount() {
```
Comment (MicahSchiewe) [inline food-truck-frontend/src/components/truck/rate_and_review/ReviewView.tsx:15]: Commented-out code

File: food-truck-frontend/src/components/truck/rate_and_review/ReviewView.tsx [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+        loadReviewById(this.props.review.reviewId).then(rev => {
+            if (rev)
+                this.setState({review: rev})
+        })
+    }*/
+
+    render() {
+        console.log(this.props.review);
```
Comment (MicahSchiewe) [inline food-truck-frontend/src/components/truck/rate_and_review/ReviewView.tsx:23]: May not want debug prints on master? May not have time to clean it, though?

File: food-truck-frontend/src/domain/Subscription.ts [hunk 12]
```diff
@@ -0,0 +1,11 @@
+import {TruckMeta} from "./Truck";
+import * as t from 'io-ts'
+
+export const SubscriptionMeta = t.type({
```
Comment (MicahSchiewe) [inline food-truck-frontend/src/domain/Subscription.ts:4]: I guess we don't need to track who's subscription it is?

File: food-truck-frontend/src/pages/account.tsx [hunk 13]
```diff
@@ -21,86 +19,101 @@ class AccountPageComponent extends React.Component<PageProps, PageState> {
...
+                            <Select label="New Privacy Setting" name="newPrivacySetting" defaultValue={this.state.user.privacySetting}>
+                                <MenuItem value="PUBLIC">Public</MenuItem>
+                                <MenuItem value="USERS_ONLY">Users Only</MenuItem>
+                                <MenuItem value="PRIVATE">Private</MenuItem>
+                            </Select>
+                            <FormHelperText>New Privacy Setting</FormHelperText>
+                        </FormControl>
+                        <Grid container direction="row">
+                            <p>Owner account?</p>
+                            <Switch name="newOwnerStatus" checked={this.state.user.owner}/>
```
Comment (MicahSchiewe) [inline food-truck-frontend/src/pages/account.tsx:58]: So we're letting them switch accounts user -> owner now? Do we block switching owner -> user? That could cause... problems.

File: food-truck-frontend/src/pages/account.tsx [hunk 14]
```diff
@@ -21,86 +19,101 @@ class AccountPageComponent extends React.Component<PageProps, PageState> {
...
-                    })
-                })
-                .catch(err => {
-                    this.setState({
-                        done: true,
-                        error: "Failed to connect to server."
-                    })
-                });
+            const user = await loadUserById(curUser.userID);
+            console.log(user);
```
Comment (MicahSchiewe) [inline food-truck-frontend/src/pages/account.tsx:112]: Debug print

File: food-truck-frontend/src/pages/interactive-map.tsx [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+    getNearbyTruckLocations(DEFAULT_ERR_RESP)
+      .then(locs => {
+        if (locs) {
+          setLocations(locs);
+          setReady(true);
+        }
+      });
+    });
+    return(
+        ready && <TruckLocationMapComponent locations={locations} />
```
Comment (MicahSchiewe) [inline food-truck-frontend/src/pages/interactive-map.tsx:20]: I'd like to ask we switch this to `ready? <map> : "Please wait"`. I can see how the operator precedence works after double-taking, but it's not as clear as the ternary operator.

File: food-truck-frontend/src/pages/user/[username].tsx [hunk 16]
```diff
@@ -13,7 +13,7 @@ function renderError(){
 
 
 function CustomerDetailPage(){
-    const router: NextRouter = useRouter();
+    /*const router: NextRouter = useRouter();
```
Comment (MicahSchiewe) [inline food-truck-frontend/src/pages/user/[username].tsx:16]: Commented-out code

**Review by MicahSchiewe — APPROVED** [review]
I'm pretty sure everything looked fine, but I can't read all of my comments, so check those to see if there's anything unusually important.

All of the 'commented-out code' and 'debug print' comments can be ignored in the interest of time, we could go through and do one last cleaning before submission tomorrow night if need be.

File: food-truck-frontend/src/components/UserDetails.tsx [hunk 17]
```diff
@@ -1,49 +1,96 @@
...
     render() {
-        if(this.state.data.length == 0){
-            return(
-                <div>
-                    <h1>User Doesn't Exist</h1>
-                </div>
-            );
-        }
+        if (this.state.err)
+            return <p>{this.state.err}</p>
```
Comment (nnazo) [inline food-truck-frontend/src/components/UserDetails.tsx:57]: If I can get around to it i'll try to update this page to use material-UI stuff but for now it's fine

File: food-truck-frontend/src/pages/_app.js [hunk 18]
```diff
@@ -22,16 +23,20 @@ const FoodTruckApp = ({Component, pageProps}) => {
...
-      </FoodTruckThemeProvider>
-    </Provider>
+      <Provider store={store}>
+          <Head>
+              <title>Stacked Trucks</title>
+              <meta name="viewport" content="minimum-scale=1, initial-scale=1, width=device-width"/>
+          </Head>
+
+
+          <MuiThemeProvider>
```
Comment (nnazo) [inline food-truck-frontend/src/pages/_app.js:33]: we're gonna end up with two of these with my PR
we'll need to fix on merging them together

**Review by nnazo — APPROVED** [review]
LGTM

File: food-truck-frontend/src/api/RateReviewApi.ts (hunk 2 above)
Comment (MicahSchiewe) [reply food-truck-frontend/src/api/RateReviewApi.ts:16]: After some research, I've concluded this is handled correctly internally, no matter how many promises wrap each other. Please disregard this comment.

**Pull request merged**